File: Dynamic_Programming/october/7_Oct_2019_nth_Catalan_number/7_Oct_2019_nth_Catalan_number_dynamic.cpp
//dynamic solution to catalan numbers

#include<bits/stdc++.h>
using namespace std;

int catalan(int n){
    int table[n+1];
    //base case

    table[0] = table[1] = 1;

    //fill in bottom up manner

    for(int i = 2; i <= n; i++){
        table[i] = 0;
        for(int j = 0; j < i; j++){
            table[i] += table[j] * table[i-j-1];
        }
    }

    return table[n];
}

//driver program
int main(){
    for (int i = 0; i< 10; i++){
        cout << catalan(i) << " ";
    }
    return 0;
}

File: Dynamic_Programming/october/7_Oct_2019_nth_Catalan_number/7_Oct_2019_nth_Catalan_number_dynamic.py
#dynamic implementation of catalna

def catalan(n):
    if(n == 0 or n == 1):
        return 1
    
    table = [0 for i in range(n+1)]

    table[0] = 1
    table[1] = 1

    for i in range(2,n+1):
        table[i] = 0
        for j in range(i):
            table[i] = table[i] + table[j] * table[i-j-1]
    
    return table[n]

#driver code
for i in range(11):
    print(" {}".format(catalan(i)))